### PR TITLE
VP-1419 migration to v3

### PR DESCRIPTION
**Pull request by AliveMen**
Initial commit

### CODE REVIEW

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs [hunk 1]
```diff
@@ -75,48 +78,49 @@ public IHttpActionResult GetTagsCount(string id)
...
         }
 
         /// <summary>
         /// POST: api/personalization/search
         /// </summary>
         [HttpPost]
         [Route("search")]
-        [ResponseType(typeof(GenericSearchResult<TaggedItem>))]
-        public IHttpActionResult Search(TaggedItemSearchCriteria criteria)
+        [ProducesResponseType(typeof(GenericSearchResult<TaggedItem>), StatusCodes.Status200OK)]
```
Comment (yecli) [inline src/VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs]: Let's add a specific class for `GenericSearchResult<TaggedItem>` to have good swagger doc (means change search service interface too).

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/Localizations/en.VirtoCommerce.CatalogPersonalization.json [hunk 2]
```diff
@@ -35,5 +35,21 @@
...
+        "TagsInheritancePolicy": {
+          "description": "DownTree - propagates all tags assigned to categories for all their descendants. UpTree - propagates tags up to hierarchy from descendants to parents",
+          "title": "Selected policy for tags propagation (requires restart)"
+        },
+        "CronExpression": {
+          "description": "x minutes use '0/x * * * *', x hourse use '0 */x * * *',  you can set minutes also '15 */x * * *', x days '0 0 */x * *'   ",
+          "title": "Job cron expression for synchronize the outlines of tagged entities. Valid only when the UpTree propagation policy is selected. (requires restart)"
+        }
+      }
+     }
```
Comment (yecli) [inline src/VirtoCommerce.CatalogPersonalizationModule.Web/Localizations/en.VirtoCommerce.CatalogPersonalization.json]: Maybe permission localization is missing

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/Module.cs [hunk 3]
```diff
@@ -51,8 +52,8 @@ public void Initialize(IServiceCollection serviceCollection)
             serviceCollection.AddSingleton<ProductTaggedItemDocumentBuilder>();
             serviceCollection.AddSingleton<CategoryTaggedItemDocumentBuilder>();
             
-            serviceCollection.AddTransient<ISearchRequestBuilder, ProductSearchUserGroupsRequestBuilder>();
-            serviceCollection.AddTransient<ISearchRequestBuilder, CategorySearchUserGroupsRequestBuilder>();
+            //serviceCollection.AddTransient<ISearchRequestBuilder, ProductSearchUserGroupsRequestBuilder>();
```
Comment (yecli) [inline src/VirtoCommerce.CatalogPersonalizationModule.Web/Module.cs]: Please remove comments

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/Properties/launchSettings.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (yecli) [inline src/VirtoCommerce.CatalogPersonalizationModule.Web/Properties/launchSettings.json]: Do we need this file?

File: src/VirtoCommerce.CatalogPersonalizationModule.Data/Repositories/PersonalizationDbContext.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        protected PersonalizationDbContext(DbContextOptions options) : base(options)
+        {
+        }
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+
+            modelBuilder.Entity<TaggedItemEntity>().ToTable("TaggedItem").HasKey(x => x.Id);
```
Comment (yecli) [inline src/VirtoCommerce.CatalogPersonalizationModule.Data/Repositories/PersonalizationDbContext.cs]: Aren't there any foreign keys between the tables, cascade deletions?

File: tests/VirtoCommerce.CatalogPersonalizationModule.Test/VirtoCommerce.CatalogPersonalizationModule.Test.csproj [hunk 6]
```diff
@@ -0,0 +1,21 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+      <noWarn>1591</noWarn>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.4.0" />
+    <PackageReference Include="MockQueryable.Moq" Version="3.0.2" />
+    <PackageReference Include="Moq" Version="4.13.1" />
+    <PackageReference Include="VirtoCommerce.ExportModule.CsvProvider" Version="3.0.0-rc.2" />
```
Comment (yecli) [inline tests/VirtoCommerce.CatalogPersonalizationModule.Test/VirtoCommerce.CatalogPersonalizationModule.Test.csproj]: Why do we need export module for the tests?

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/module.manifest [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+<module>
+    <id>VirtoCommerce.CatalogPersonalization</id>
+    <version>3.0.0</version>
+    <version-tag>rc.3</version-tag>
+    <platformVersion>3.0.0</platformVersion>
+    <dependencies>
+        <dependency id="VirtoCommerce.Core" version="3.0.0-rc.3" />
+        <dependency id="VirtoCommerce.Customer" version="3.0.0-rc.3" />
+        <dependency id="VirtoCommerce.Catalog" version="3.0.0-rc.3" />
+    </dependencies>
```
Comment (yecli) [inline src/VirtoCommerce.CatalogPersonalizationModule.Web/module.manifest]: Probably we need search module in dependencies

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/module.manifest [hunk 8]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<module>
+    <id>VirtoCommerce.CatalogPersonalization</id>
+    <version>3.0.0</version>
+    <version-tag>rc.3</version-tag>
+    <platformVersion>3.0.0</platformVersion>
+    <dependencies>
+        <dependency id="VirtoCommerce.Core" version="3.0.0-rc.3" />
+        <dependency id="VirtoCommerce.Customer" version="3.0.0-rc.3" />
```
Comment (yecli) [inline src/VirtoCommerce.CatalogPersonalizationModule.Web/module.manifest]: Do we need customer module reference?

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs (hunk 1 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs]: fixed

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/Localizations/en.VirtoCommerce.CatalogPersonalization.json (hunk 2 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogPersonalizationModule.Web/Localizations/en.VirtoCommerce.CatalogPersonalization.json]: added

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/Module.cs (hunk 3 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogPersonalizationModule.Web/Module.cs]: removed

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/Properties/launchSettings.json (hunk 4 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogPersonalizationModule.Web/Properties/launchSettings.json]: deleted

File: tests/VirtoCommerce.CatalogPersonalizationModule.Test/VirtoCommerce.CatalogPersonalizationModule.Test.csproj (hunk 6 above)
Comment (AliveMen) [reply tests/VirtoCommerce.CatalogPersonalizationModule.Test/VirtoCommerce.CatalogPersonalizationModule.Test.csproj]: removed

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/module.manifest (hunk 7 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogPersonalizationModule.Web/module.manifest]: added

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/module.manifest (hunk 8 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogPersonalizationModule.Web/module.manifest]: removed

File: tests/VirtoCommerce.CatalogPersonalizationModule.Test/VirtoCommerce.CatalogPersonalizationModule.Test.csproj [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+      <noWarn>1591</noWarn>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.4.0" />
+    <PackageReference Include="MockQueryable.Moq" Version="3.0.2" />
+    <PackageReference Include="Moq" Version="4.13.1" />
+    <PackageReference Include="VirtoCommerce.ExportModule.Data" Version="3.0.0-rc.2" />
+    <PackageReference Include="VirtoCommerce.ExportModule.JsonProvider" Version="3.0.0-rc.2" />
```
Comment (yecli) [inline tests/VirtoCommerce.CatalogPersonalizationModule.Test/VirtoCommerce.CatalogPersonalizationModule.Test.csproj]: Seems we do not need export packages here at all, do we?

File: src/VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs [hunk 10]
```diff
@@ -29,55 +28,9 @@ public class PersonalizationService : ServiceBase, ITaggedItemService, ITaggedIt
...
-                CommitChanges(repository);
-                pkMap.ResolvePrimaryKeys();
-            }
-        }
-
-        public GenericSearchResult<TaggedItem> SearchTaggedItems(TaggedItemSearchCriteria criteria)
+        public async Task<GenericSearchResult<TaggedItem>> SearchTaggedItemsAsync(TaggedItemSearchCriteria criteria)
         {
-            var retVal = new GenericSearchResult<TaggedItem>();
+            var result = new GenericSearchResult<TaggedItem>();
```
Comment (yecli) [inline src/VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs]: Please use caching for search and all Get operations. 
In save and delete you need to `PersonalizationCacheRegion.ExpireRegion()`
[Example](https://github.com/VirtoCommerce/vc-module-marketing/blob/release/3.0.0/src/VirtoCommerce.MarketingModule.Data/Services/CouponService.cs#L37)

Comment (AliveMen) [reply src/VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs]: done

File: src/VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs [hunk 11]
```diff
@@ -17,99 +20,108 @@ public class PersonalizationService : ITaggedItemService, ITaggedItemSearchServi
...
 
         public async Task<GenericSearchResult<TaggedItem>> SearchTaggedItemsAsync(TaggedItemSearchCriteria criteria)
         {
-            var result = new GenericSearchResult<TaggedItem>();
 
+            var cacheKey = CacheKey.With(GetType(), nameof(SearchTaggedItemsAsync), criteria.GetCacheKey());
+            return await _platformMemoryCache.GetOrCreateExclusiveAsync(cacheKey,
+                async (cacheEntry) =>
+                {
+                    cacheEntry.AddExpirationToken(PersonalizationCacheRegion.CreateChangeToken());
```
Comment (yecli) [inline src/VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs]: Need to eliminate `GenericSearchResult<TaggedItem>` in favor of `TaggedItemSearchResult`.
Controller uses it.

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs [hunk 12]
```diff
@@ -75,48 +78,49 @@ public IHttpActionResult GetTagsCount(string id)
...
+            return Ok(count);
         }
 
         /// <summary>
         /// PUT: api/personalization/taggeditem
         /// </summary>
         [HttpPut]
         [Route("taggeditem")]
-        public IHttpActionResult UpdateTaggedItem(TaggedItem taggedItem)
+        [ProducesResponseType(typeof(void), StatusCodes.Status204NoContent)]
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply src/VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs]: changed to NoContent

File: src/VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs [hunk 13]
```diff
@@ -141,9 +145,9 @@ public IHttpActionResult RunOutlinesSynchronization()
         /// PUT: api/personalization/outlines/synchronization/cancel
         /// </summary>
         [HttpPost]
-        [ResponseType(typeof(void))]
         [Route("outlines/synchronization/cancel")]
-        public IHttpActionResult CancelSynchronization([FromBody]TaggedItemOutlinesSynchronizationRequest cancellationRequest)
+        [ProducesResponseType(typeof(void), StatusCodes.Status204NoContent)]
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply src/VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs]: changed to NoContent

**Pull request merged**